test/end-to-end/shared.sh

```diff
@@ -82,3 +73,19 @@ cleanup_supervisor() {
     hab sup term
     sed -e 's/^/TEST_SUP_LOG>>> /' "${sup_log_file}" >&2
 }
+
+# Waits up to 10 seconds for a Supervisor is listening on its default
+# control gateway port. A Supervisor process may be running, but we
+# can't really interact with it until the gateway is up.
+#
+# NOTE: This currently assumes the test is the only thing running a
+# Supervisor.
+wait_for_control_gateway() {
+    # TODO (CM): Ideally would use -z, but that appears to cause
+    # issues with the control gateway in our Docker-based testing
+    # pipeline... something about the peer address of the connection
+    # not being available.
+    #
+    # Also, this assumes GNU netcat, installed via Habitat.
+    timeout --foreground 10 sh -c 'until netcat -vv --close localhost 9632; do sleep 1; done'
```

Comment:
Yeah, it appears that `-z` causes the client to close the socket before the gateway can send a response so it panics sending on a socket that is now closed.

I think what you have will be OK - however in my testing I had to increase the timeout from 10s.
I'm seeing it take 10s of silence on the socket before the ctl_gateway closes the connection:

```
time timeout --foreground 20 sh -c 'until /hab/pkgs/core/netcat/0.7.1/20190117175043/bin/netcat -vv --close localhost 9632; do sleep 1; done'
localhost [127.0.0.1] 9632 open
[2019-11-21T16:43:42Z DEBUG tokio_core::reactor] loop poll - 110.753428303s
[2019-11-21T16:43:42Z DEBUG tokio_core::reactor] loop time - Instant { tv_sec: 263228, tv_nsec: 141580969 }
[2019-11-21T16:43:42Z DEBUG tokio_core::reactor] loop process, 204.52µs
[2019-11-21T16:43:42Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2019-11-21T16:43:42Z DEBUG tokio_reactor] adding I/O source: 25165825
[2019-11-21T16:43:42Z DEBUG tokio_reactor::registration] scheduling Read for: 1
[2019-11-21T16:43:42Z DEBUG tokio_core::reactor] loop poll - 248.889µs
[2019-11-21T16:43:42Z DEBUG tokio_core::reactor] loop time - Instant { tv_sec: 263228, tv_nsec: 142420718 }
[2019-11-21T16:43:42Z DEBUG tokio_core::reactor] loop process, 144.841µs
[2019-11-21T16:43:50Z DEBUG tokio_core::reactor] loop poll - 8.526368668s
[2019-11-21T16:43:50Z DEBUG tokio_core::reactor] loop time - Instant { tv_sec: 263236, tv_nsec: 669020147 }
[2019-11-21T16:43:50Z DEBUG tokio_core::reactor] loop process, 221.268µs
[2019-11-21T16:43:52Z DEBUG tokio_core::reactor] loop poll - 1.471664011s
[2019-11-21T16:43:52Z DEBUG tokio_core::reactor] loop time - Instant { tv_sec: 263238, tv_nsec: 141015699 }
[2019-11-21T16:43:52Z DEBUG tokio_core::reactor] loop process, 215.255µs
[2019-11-21T16:43:52Z DEBUG tokio_reactor::registration] scheduling Read for: 1
[2019-11-21T16:43:52Z DEBUG tokio_reactor] dropping I/O source: 1
Total received bytes: 0
Total sent bytes: 0
[2019-11-21T16:43:52Z DEBUG habitat_sup::ctl_gateway::server] DISCONNECTED from V4(127.0.0.1:34036) with result Err(Io(Custom { kind: TimedOut, error: "client timed out" }))
[2019-11-21T16:43:52Z DEBUG tokio_core::reactor] loop poll - 2.117066ms
[2019-11-21T16:43:52Z DEBUG tokio_core::reactor] loop time - Instant { tv_sec: 263238, tv_nsec: 143459422 }
[2019-11-21T16:43:52Z DEBUG tokio_core::reactor] loop process, 134.032µs
[2019-11-21T16:43:52Z DEBUG tokio_core::reactor] loop poll - 15.918µs

real	0m10.005s
user	0m0.003s
sys	0m0.000s
```

That being the case, perhaps increase to 20s?

Reply:
:thinking: I can't seem to find this timeout value in the code anywhere.

The panic isn't on sending a response; it's actually in accessing the peer address on the connection. Interestingly, it seems the only reason we access the address is to create that `DISCONNECTED` log message :frowning: 

We should probably remove the `expect` that's triggering this panic, and replace this `map` with a `filter_map`, returning `None` if we can't get a `peer_addr`: https://github.com/habitat-sh/habitat/blob/634c7f1523af34e96b0615162c131feef6ee0677/components/sup/src/ctl_gateway/server.rs#L484-L492

In the meantime, we could increase the timeout... or just sleep for a few seconds :man_facepalming:

Reply:
@jeremymv2 It looks like I can actually _reduce_ the `netcat` operation timeout without ill effect... I think what you're seeing is only related to the processing the supervisor itself is doing in response. It's perhaps not great, but I don't think it's causing the tests to take longer than they need to.